graf2d/gpadv7/v7/inc/ROOT/RDrawingAttr.hxx

```diff
@@ -292,6 +293,200 @@ public:
+      std::string GetString() const final { return v; }
+   };
+
+
+   using Map_t = std::unordered_map<std::string, std::unique_ptr<Value_t>>;
+
+   struct Record_t {
+      std::string type;             ///<! drawable type, not stored in the root file, must be initialized
+      std::string user_class;       ///<  user defined drawable class, can later go inside map
+      Map_t map;                    ///<  JSON_object
+      Map_t *defaults{nullptr};     ///<! default values for server
+
+      Record_t() = default;
+
+      Record_t(const Record_t &src) = delete;
+      Record_t& operator=(const Record_t &src) = delete;
+   };
+
+private:
+   Record_t                  *fContIO{nullptr};    ///  used in IO while shared_ptr not yet supported, JSON_object
+   std::shared_ptr<Record_t>  fCont;               ///<! container itself
+
+   auto *MakeContainer()
+   {
+      if (fContIO && !fCont)
+         fCont.reset(fContIO);
+
+      if (!fCont)
+         fCont = std::make_shared<Record_t>();
+
+      return fContIO = fCont.get();
+   }
+
+   const auto *GetContainer() const
+   {
+      if (!fContIO)
+         const_cast<RDrawableAttributes*>(this)->fContIO = fCont.get();
+      else if (!fCont)
+         const_cast<RDrawableAttributes*>(this)->fCont.reset(fContIO);
+
+      return fContIO;
+   }
+
+public:
+
+   RDrawableAttributes() = default;
+
+   RDrawableAttributes(const std::string &_type) { MakeContainer()->type = _type; }
+
+   ~RDrawableAttributes() { Clear(); }
+
+   std::weak_ptr<Record_t> Make()
+   {
+      MakeContainer();
+      return fCont;
+   }
+
+   std::weak_ptr<Record_t> Get() const
+   {
+      GetContainer();
+      return fCont;
+   }
+
+   void Clear()
+   {
+      // special case when container was read by I/O but not yet assigned to
+      if (fContIO && !fCont)
+         delete fContIO;
+
+      fContIO = nullptr;
+      fCont.reset();
+   }
+
+};
+
+//////////////////////////////////////////////////////////////////////////
+
+class RStyleNew {
+public:
+
+   struct Block_t {
+      std::string selector;
+      RDrawableAttributes::Map_t map; ///<   JSON_object
+      Block_t() = default;
+      Block_t(const std::string &_selector) : selector(_selector) {}
+
+      Block_t(const Block_t &) {} // dummy, should not be used, but appears in dictionary
+      Block_t& operator=(const Block_t &) = delete;
+   };
+
+   const RDrawableAttributes::Value_t *Eval(const std::string &type, const std::string &user_class, const std::string &field) const;
+
+   RDrawableAttributes::Map_t &AddBlock(const std::string &selector)
+   {
+      fBlocks.emplace_back(selector);
+      return fBlocks.back().map;
+   }
+
+private:
+   std::list<Block_t> fBlocks;  // use std::list to avoid calling of Block_t copy constructor
+};
+
+
+/** Access to drawable attributes, never should be stored */
+class RAttributesVisitor {
+
+   mutable std::weak_ptr<RDrawableAttributes::Record_t> fWeak;     ///<! weak pointer on container
```

Comment:
`mutable` makes it very hard to reason about thread-saftety behavior (breaks "const == thread-safe").

Can we have const and non-const methods instead?

Reply:
Seems to be, `mutable` not required for `weak_ptr` - I removed it now. 
`shared_ptr` and `bool` flag only used as temporary values.
One can remove them (keeping only `weak_ptr`), but then one need to create `shared_ptr` for every attributes access - which could be a penalty for complex objects with multiple access operations for same attributes.